File: 0078-subsets/0078-subsets.java
class Solution {
    public void solve(int[] nums,List<Integer>output,List<List<Integer>> ans,int ind){
        if(ind>=nums.length){
            ans.add(new ArrayList<>(output));
            return;
        }
        output.add(nums[ind]);
        solve(nums,output,ans,ind+1);
        if(output.size()>0)
         output.remove(output.size()-1);
        solve(nums,output,ans,ind+1);
        
    }

    public List<List<Integer>> subsets(int[] nums) {
        List<Integer> output=new ArrayList<>();
        List<List<Integer>> ans=new ArrayList<>();
        solve(nums,output,ans,0);
        return ans;
    }
}

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void solve(vector<int>&nums,vector<int>&output,vector<vector<int>>&ans,int ind){
        if(ind>=nums.size()){
            ans.push_back(output);
            return;
        }
        // include
        output.push_back(nums[ind]);
        solve(nums,output,ans,ind+1);
        output.pop_back();
        solve(nums,output,ans,ind+1);
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int>output;
        vector<vector<int>>ans;
        solve(nums,output,ans,0);
        return ans;
    }
};